micrometer-core/src/main/java/io/micrometer/core/instrument/binder/httpcomponents/AbstractPoolingHttpClientConnectionManagerMetricsBinder.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 VMware, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.httpcomponents;
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.lang.NonNull;
+import org.apache.http.conn.routing.HttpRoute;
+import org.apache.http.pool.ConnPoolControl;
+import org.apache.http.pool.PoolStats;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class which collects total metrics including {@link HttpRoute} state from {@link ConnPoolControl}, for example {@link org.apache.http.impl.conn.PoolingHttpClientConnectionManager}
+ * for synchronous HTTP clients or {@link org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager} for asynchronous HTTP clients.
+ * <p>
+ * It monitors the overall and per route connection pool state.
+ *
+ * @author Mikhail Yakimchenko (mikhailyakimchenko@gmail.com)
+ * @since 1.5.0
+ */
+abstract class AbstractPoolingHttpClientConnectionManagerMetricsBinder<T extends ConnPoolControl<HttpRoute>> extends
+        ConnPoolControlMetricsBinder<HttpRoute> {
+
+    @SuppressWarnings("unchecked")
+    private final T connectionManager = (T) connPoolControl;
+    private final ConcurrentMap<String, PoolStats> poolStatsByHost = new ConcurrentHashMap<>();
+
+    public AbstractPoolingHttpClientConnectionManagerMetricsBinder(T connectionManager, String name, String... tags) {
+        this(connectionManager, name, Tags.of(tags));
+    }
+
+    public AbstractPoolingHttpClientConnectionManagerMetricsBinder(T connectionManager, String name, Iterable<Tag> tags) {
+        super(connectionManager, name, tags);
+    }
+
+    protected abstract Collection<HttpRoute> getRoutes(T connectionManager);
+
+    @Override
+    public void bindTo(@NonNull MeterRegistry registry) {
+        super.bindTo(registry);
+        updateRoutesMetrics(registry);
+    }
+
+    /**
+     * Adds metrics for newly created {@link HttpRoute} to the given meter registry.
+     */
+    public void updateRoutesMetrics(MeterRegistry registry) {
+        for (HttpRoute route : getRoutes(connectionManager)) {
```

Comment:
👋 I come across this PR trying to resolve #1616 as I have similar requirements as well. Great initiative!

However, I wonder how this piece will work as the gauge per route is registered at "init" time and won't be updated going forward. If I understand the problem correctly, we need to add a gauge on a route created event but it's buried deep into the apache httpclient implementation...

Reply:
Hi, Alan! Your comprehension of the problem is absolutely correct. Thanks for having a look 👍 

I didn't find out another way of exposing the routes by using contract provided by`PoolingHttpClientConnectionManager` class. So I've decided to move the responsibility of scheduling exposed method `updateRoutesMetrics` to the client side. That's not obvious, so I need to expand docs at least.

Potentially the problem can be solved by registering another gauge for that case. I'm not sure if it's a good idea because each call of `getRoutes` method cause acquiring the lock and creation of another `HashSet` object.